src/solvers/refinement/string_constraint_generator.h

```diff
@@ -0,0 +1,375 @@
+/*******************************************************************\
+
+Module: Generates string constraints to link results from string functions
+        with their arguments. This is inspired by the PASS paper at HVC'13:
+        "PASS: String Solving with Parameterized Array and Interval Automaton"
+        by Guodong Li and Indradeep Ghosh, which gives examples of constraints
+        for several functions.
+
+Author: Romain Brenguier, romain.brenguier@diffblue.com
+
+\*******************************************************************/
+
+#ifndef CPROVER_SOLVERS_REFINEMENT_STRING_CONSTRAINT_GENERATOR_H
+#define CPROVER_SOLVERS_REFINEMENT_STRING_CONSTRAINT_GENERATOR_H
+
+#include <solvers/refinement/string_expr.h>
+#include <solvers/refinement/string_constraint.h>
+
+class string_constraint_generatort
+{
+public:
+  // This module keeps a list of axioms. It has methods which generate
+  // string constraints for different string funcitons and add them
+  // to the axiom list.
+
+  string_constraint_generatort(): mode(ID_unknown) { }
+
+  void set_mode(irep_idt _mode)
+  {
+    // only C and java modes supported
+    assert((_mode==ID_java) || (_mode==ID_C));
+    mode=_mode;
+  }
+
+  inline irep_idt &get_mode() { return mode; }
+
+  unsignedbv_typet get_char_type() const;
+  inline signedbv_typet get_index_type() const
+  { return refined_string_typet::index_type(); }
+
+  // Axioms are of three kinds: universally quantified string constraint,
+  // not contains string constraints and simple formulas.
+  std::vector<exprt> axioms;
+
+  // Boolean symbols for the results of some string functions
+  std::vector<symbol_exprt> boolean_symbols;
+
+  // Symbols used in existential quantifications
+  std::vector<symbol_exprt> index_symbols;
```

Comment:
No good reason, I've replaced them by lists, which are enough for our purpose https://github.com/diffblue/cbmc/pull/374/commits/3368a3d57c9decb7cdb10f70a7edff1b2cc101aa